Reprise of #20: serde support

Uriopass opened the pull request:
Adds serde support through a serde feature. I do not know why the author's old PR was closed since it all seems to work perfectly. Only changed std:: to core:: to support no-std and other small things like readme update and travis serde test.  

Fixes #41

kvark commented on src/vector.rs:
```diff
@@ -41,6 +41,48 @@ macro_rules! vec {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+                where S: ::serde::Serializer
+            {
+                use ::serde::ser::SerializeSeq;
```
why aren't we piggy-backing on the `AsRef` implementation? E.g. this function would look like `self.as_ref().serialize(serializer)`

Uriopass replied:
:heavy_check_mark:

Uriopass commented on Cargo.toml:
```diff
@@ -16,3 +16,7 @@ documentation = "https://docs.rs/mint"
 keywords = ["math"]
 
 [dependencies]
+serde = { version = "1.0", optional = true, features = ["derive"] }
```
From [serde's documentation](https://serde.rs/derive.html): 
```
These derives require a Rust compiler version 1.31 or newer.
Add serde = { version = "1.0", features = ["derive"] } as a dependency in Cargo.toml.
```

Uriopass replied:
I also compiled with `version = "=1.0.0"` and it passed all tests and compiled successfully.

kvark replied:
alright, thanks for checking! I stand corrected

kvark commented on src/matrix.rs:
```diff
@@ -51,6 +51,45 @@ macro_rules! matrix {
+        impl<'de, T> ::serde::Deserialize<'de> for $name<T>
+            where T: ::serde::Deserialize<'de>
+        {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
```
why don't we piggy-back on fixed-size slices here as well? something like:
```rust
[[T; $inner]; $outer].deserialize(D).map($name<T>::from)
```

Uriopass replied:
I thought about that, I'm just not sure if the extra conversion is worth it ? Maybe Rust will optimize it to a no-op.

kvark replied:
We can check if it does optimize this by looking at the assembly, even through play.rust-lang.org

Uriopass commented:
Rebased, removed dependency on `serde/derive`, added tests for quaternions/euler angles, used simpler serialize/deserialize code piggybacking on AsRef and From, except for EulerAngles which doesn't implement AsRef<[T; 3]>.

kvark reviewed (approved):
Thank you, amazing work!
